contactsummary/gulpfile.js

```diff
@@ -9,110 +9,220 @@ var exec = require('child_process').exec;
+   * - fullPath: The path + the sub path specified in the build file. Ex:
+   *   /root/path/to/ext/js/angular/src/shared'
+   */
+  function getExtensionsFromBuildContent (buildFileContent) {
+    var extensions = [];
+    var extensionPathRegExp = /((%{([\w._-]+)})(.*))/g;
+
+    // linter doesn't like assignments as condition.
+    while (true) {
+      var matches = extensionPathRegExp.exec(buildFileContent);
+
+      if (!matches) break;
+
+      var extension = {
+        name: matches[3],
+        placeholder: matches[2],
+        subPath: matches[4]
+      };
+      var extensionPath = getExtensionPathByName(extension.name);
+
+      extensions.push({
+        placeholder: extension.placeholder,
+        path: extensionPath,
+        fullPath: path.join(extensionPath, extension.subPath)
+      });
     }
 
-    return {
-
-        /**
-         * Runs all the tests
-         */
-        all: function () {
-            runServer('karma.conf.js');
-        },
-
-        /**
-         * Runs the tests for a specific source file
-         *
-         * Looks for a test file (*_test.js) in `test/`, using the same path
-         * of the source file in `src/contactsummary/`
-         *   i.e. src/contactsummary/models/model.js -> test/models/model_test.js
-         *
-         * @throw {Error}
-         */
-        for: function (srcFile) {
-            var srcFileNoExt = path.basename(srcFile, path.extname(srcFile));
-            var testFile = srcFile
-                .replace('src/contactsummary/', 'test/')
-                .replace(srcFileNoExt + '.js', srcFileNoExt + '_test.js');
-
-            fs.statSync(testFile).isFile() && this.single(testFile);
-        },
-
-        /**
-         * Runs a single test file
-         *
-         * It passes to the karma server a temporary config file
-         * which is deleted once the test has been run
-         *
-         * @param {string} testFile - The full path of a test file
-         */
-        single: function (testFile) {
-            var configFile = 'karma.' + path.basename(testFile, path.extname(testFile))  + '.conf.temp.js';
-
-            gulp.src(__dirname + '/js/karma.conf.js')
-                .pipe(replace('*_test.js', path.basename(testFile)))
-                .pipe(rename(configFile))
-                .pipe(gulp.dest(__dirname + '/js'))
-                .on('end', function () {
-                    runServer(configFile, function () {
-                        gulp.src(__dirname + '/js/' + configFile, { read: false }).pipe(clean());
-                    });
-                });
-        }
-    };
+    return extensions;
+  }
+
+  /**
+   * Returns watch paths for JS files that are part of the extensions specified
+   * in the build file.
+   *
+   * @return {Array}
+   */
+  function getExtensionWatchPaths () {
+    var buildFileContent = getBuidFileContent();
+    var requiredExtensions = getExtensionsFromBuildContent(buildFileContent);
```

Comment:
I was trying to avoid that since this is not necessary for sass tasks, etc. I could do as you say, or I could return a constructor function, something like this:

```js
function buildFileManager () {
  var externalExtensions;

  this.createTempFile = createTempFile;
  this.getExtensionWatchPaths = getExtensionWatchPaths;

  (function init () {
    initExternalExtensions()
  })();

  // definitions
}
```

and use it like this:

```js
new buildFileManager().createTempFile()
```